zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/package.scala

```diff
@@ -112,7 +138,43 @@ package object leader {
           new CustomLeaderLock(lockName, retryPolicy, deleteLockOnRelease, configmaps)
         )) >>> fromLock
 
-    private def logLeaderElectionFailure[E](
+    /** Lease based leader election implementation
+      *
+      * The leadership is not guaranteed to be held forever, the effect executed in runAsLeader
+      * may be interrupted. It is recommended to retry runAsLeader in these cases to try to reacquire the lease.
+      *
+      * This is a reimplementation of the Go leaderelection package:
+      * https://github.com/kubernetes/client-go/blob/master/tools/leaderelection/leaderelection.go
+      *
+      * @param lockName Name of the lease resource
+      * @param leaseDuration Duration non-leader candidates must wait before acquiring leadership. This is measured
+      *                      against the time of the last observed change.
+      * @param renewTimeout The maximum time a leader is allowed to try to renew its lease before giving up
+      * @param retryPeriod  Retry period for acquiring and renewing the lease
+      */
+    def leaseLock(
+      lockName: String,
+      leaseDuration: Duration = 15.seconds,
+      renewTimeout: Duration = 10.seconds,
+      retryPeriod: Duration = 2.seconds
+    ): ZLayer[Random with ContextInfo with Leases, Nothing, LeaderElection] =
+      (for {
+        selfInfo       <- ZIO.service[ContextInfo.Service]
+        leases         <- ZIO.service[Leases.Service]
+        random         <- ZIO.service[Random.Service]
+        leadershipLost <- Queue.bounded[Unit](1)
```

Comment:
We should either:
- make a big note that this queue must be consumed or the process will deadlock
- make this a sliding queue

Reply:
I also don't see where this queue is exposed to the users of the library so is that an oversight?

Reply:
👍 it is not exposed to users though

Reply:
It is used in the `LiveTemporary` 's `runAsLeader`. I did not want to break the existing API (for no good reason actually :D )